tests/tests/test_mtls.py

```diff
@@ -76,7 +76,59 @@ def make_script_artifact(artifact_name, device_type, output_path):
 
 
 class TestClientMTLSEnterprise:
-    def common_test_mtls_enterprise(self, setup_ent_mtls, algorithm=None):
+    def get_key_uri_hsm(self, ssl_engine_id, device):
+        algorithm = "rsa"
+        tmpdir = tempfile.mkdtemp()
+        key = f"/var/lib/mender/client.1.{algorithm}.key"
+        pin = "0001"
+        script = f"""\
+#!/bin/bash
+echo "module: /usr/lib/softhsm/libsofthsm2.so" > /usr/share/p11-kit/modules/softhsm2.module
+mkdir -p /softhsm/tokens
+echo "directories.tokendir = /softhsm/tokens" > /softhsm/softhsm2.conf
+export SOFTHSM2_CONF=/softhsm/softhsm2.conf;
+softhsm2-util --init-token --free --label unittoken1 --pin {pin} --so-pin 0002
+pkcs11-tool --module /usr/lib/softhsm/libsofthsm2.so --login --pin {pin} --write-object "{key}" --type privkey --id 0909 --label privatekey
+"""
+        initialize_hsm_script = os.path.join(tmpdir, "init-hsm.sh")
+        with open(initialize_hsm_script, "w") as fd:
+            fd.write(script)
+        device.put(
+            os.path.basename(initialize_hsm_script),
+            local_path=os.path.dirname(initialize_hsm_script),
+            remote_path="/tmp",
+        )
+        device.run("chmod 755 /tmp/" + os.path.basename(initialize_hsm_script))
+        device.run("/tmp/" + os.path.basename(initialize_hsm_script))
+        shutil.rmtree(tmpdir)
+
+        key_uri = device.run(
+            "export SOFTHSM2_CONF=/softhsm/softhsm2.conf; p11tool --login --provider=/usr/lib/softhsm/libsofthsm2.so --set-pin="
+            + pin
+            + " --list-all-privkeys | grep URL | sed -e 's/.*URL: //'"
+        ).rstrip("\n")
+        key_uri = key_uri + ";pin-value=" + pin
+        device.run("cp /etc/ssl/openssl.cnf /etc/ssl/openssl.cnf.backup")
+        device.run(
+            'echo -ne "[openssl_init]\nengines=engine_section\n\n[engine_section]\npkcs11 = pkcs11_section\n\n[pkcs11_section]\nengine_id = '
+            + ssl_engine_id
+            + '\nMODULE_PATH = /usr/lib/softhsm/libsofthsm2.so\ninit = 0\n" >> /etc/ssl/openssl.cnf'
+        )
+        device.run(
+            'sed -i.backup -e "/\[Service\]/ a Environment=SOFTHSM2_CONF=/softhsm/softhsm2.conf" /lib/systemd/system/mender-client.service'
```

Comment:
Thank you so much for this! I didn't know about the `a` command in sed, and this is something I have been looking for for long...

Reply:
that makes two of us :) I always look for it every time I want to use it